pkg/image/registryclient/alternativesources.go

```diff
@@ -0,0 +1,193 @@
+package registryclient
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+
+	"github.com/docker/distribution"
+	operatorv1alpha1 "github.com/openshift/api/operator/v1alpha1"
+	operatorv1alpha1scheme "github.com/openshift/client-go/operator/clientset/versioned/scheme"
+	operatorv1alpha1client "github.com/openshift/client-go/operator/clientset/versioned/typed/operator/v1alpha1"
+	imagereference "github.com/openshift/library-go/pkg/image/reference"
+)
+
+// AlternativeImageSources holds ImageContentSourcePolicy variables to look up image sources
+type AlternativeImageSources struct {
+	ICSPFile     string
+	ICSPList     []operatorv1alpha1.ImageContentSourcePolicy
+	ICSPClientFn func() (operatorv1alpha1client.ImageContentSourcePolicyInterface, error)
+}
+
+// WithICSP returns a Context with ImageContentSourcePolicy variables populated
+func (c *Context) WithICSP(icspFile string, icspClientFn func() (operatorv1alpha1client.ImageContentSourcePolicyInterface, error)) *Context {
+	c.ImageSources = AlternativeImageSources{
+		ICSPFile:     icspFile,
+		ICSPClientFn: icspClientFn,
+	}
+	return c
+}
+
+// addICSPsFromCluster will lookup ICSPs from cluster. Logs errors rather than returning errors. By design,
+// this function has no way of knowing whether it expects to connect to a cluster or not. That is up the the caller.
+func (a *Context) addICSPsFromCluster(icspClient operatorv1alpha1client.ImageContentSourcePolicyInterface) {
+	if icspClient == nil {
```

Comment:
Why do you need to pass client here directly, if you can easily access `a.ICSPClientFn` check if it's nil or not and act on it.